src/main/clojure/clara/rules/durability/fressian.clj

```diff
@@ -492,11 +492,13 @@
                 (.writeObject w (.-node-id ^RuleOrderedActivation c) true)
                 (.writeObject w (.-token ^RuleOrderedActivation c))
                 (.writeObject w (.-activation ^RuleOrderedActivation c))
-                (.writeInt w (.-rule-load-order ^RuleOrderedActivation c))))
+                (.writeInt w (.-rule-load-order ^RuleOrderedActivation c))
+                (.writeBoolean w (mem/use-token-identity? ^RuleOrderedActivation c))))
```

Comment:
Why do you have the type hint on c here?  Type hints are normally used so that Clojure can resolve a Java interop call to a particular function at compile time.  With protocol functions the workflow in the end is something like:

- Clojure calls the protocol function like any other Clojure function.  The function has a method like invoke(Object arg1 .. Object argN) if you look at the emitted bytecode.
- That function searches for the implementation on the type in question; see [here](https://github.com/clojure/clojure/blob/clojure-1.9.0/src/clj/clojure/core_deftype.clj#L602) if you're curious.

This will actually be a bit more performant if you invoke the JVM method directly; it will be mangled a bit for the ? I think but it will exist.  The reason to use the protocol function instead is if you extend the protocol outside the definition of the type, for example with [extend-protocol](https://clojuredocs.org/clojure.core/extend-protocol), but that isn't the case here.  If you do this you'll want the type hint.  Might be good to do for the sake of the pattern although it doesn't really matter here unless a user reuses the handler in a hotspot.  That sort of thing could matter a bit if you were doing it on a type here that is actually used though and it is an easy optimization to make.

Reply:
The hint isn't useful. I think it is just left over from when the field was being directly accessed here in the first revision of this PR (that wouldn't work though since it is a private field and that's why this change was made).

I don't think that direct interop with a protocol is going to be important here. The inline caching would likely be just about the same and this isn't a particularly "hot spot" (not even used by durability at this point and not used elsewhere since it was broken before this). Best to stay basic when it can be.

Reply:
Not a big deal either way as you say.  I'm OK with using the protocol here if we make this change, I mainly just wanted to raise awareness for cases where it might matter.